### PR TITLE
Calendar View

### PR DESCRIPTION
**Pull request by jlews**
## What's new?

- Calendar view component created
- Calendar view JSDocs written.

NOTES:
- Does not yet contain the calendar input, only the calendar view that will be used by the input and can also be used as a standalone component.
- Styling and story-booking still needs doing.

## Ticket number(s) in JIRA (if internal)

N/A - all tickets relate to the input, but this needed to be built first.

[board](https://rocketmakers.atlassian.net/jira/software/projects/HM/boards/172)

## Checklist

- [N] are your changes in Storybook?
- [N/A] are any breaking changes documented in `docs/migrating_from_oldstrong.md`?
- [Y] does _everything_ have jsdoc?
- [Y] is everything exported from index.ts?

### CODE REVIEW

File: module/src/components/calendarView/calendarView.component.tsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
+import * as React from 'react';
+
+import { use as useForm } from '../../hooks/form';
+import { Arrays } from '../../utils/arrays';
+import { ClassNames } from '../../utils/classNames';
+import { Dates } from '../../utils/dates';
+import { Maths } from '../../utils/maths';
+import { Button } from '../button';
+import { NativeSelectInput } from '../nativeSelectInput';
+import { getDayOfWeekHeadings, getDays, getMonths, getYears, ICalendarHighlight, ICalendarHighlightParsed, IDay } from './calendarView.utils';
```
Comment (patrickbenjamin1) [inline module/src/components/calendarView/calendarView.component.tsx:10]: worth splitting some of the more complex types into calendarView.types.ts for these more complex components or is that gonna be overkill?

File: module/src/utils/arrays.ts [hunk 2]
```diff
@@ -27,6 +27,36 @@ export namespace Arrays {
...
+   * @example If `["a", "b", "c", "d", "e"]` was passed with a `startFrom` index of `2`, the result would be `["c", "d", "e", "a", "b"]`
+   * @param array The array to clone and re-index.
+   * @param startFrom The current index of the new first item.
+   * @returns A new array, re-indexed with the item at the `startFrom` index now first.
+   */
+  export function reIndex<T>(array: T[], startFrom: number): T[] {
+    if (startFrom === 0) {
+      return array;
+    }
+    return [...[...array].splice(startFrom, array.length - startFrom), ...[...array].splice(0, startFrom)];
```
Comment (patrickbenjamin1) [inline module/src/utils/arrays.ts:42]: you could do this with slices and eliminate the need for the spreading

[...array.slice(startFrom), ...array.slice(0,startFrom]

File: module/src/utils/arrays.ts [hunk 3]
```diff
@@ -27,6 +27,36 @@ export namespace Arrays {
...
+  }
+
+  /**
+   * Creates a new array containing an item count that matches the passed `count`.
+   * - The items in the array are simply the index numbers.
+   * - Useful when you want to run a `map` x number of times but you don't have a specific array to loop.
```
Comment (patrickbenjamin1) [inline module/src/utils/arrays.ts:48]: For that purpose I'd probably wrap up the map as well to remove a bit of boilerplate, that you could use like

```Arrays.repeat(index => <></>)```

File: module/src/utils/dates.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { enGB } from 'date-fns/locale';
+
+/** Set of utilities and types relating to the native JS date object */
+export namespace Dates {
+  /** The root type for date properties.
+   * - Accepts a date as a native JS `Date` object.
+   * - Accepts a date string in IOS format.
+   * - Accepts a date string in a format matching an accompanying `format` string property.
+   */
+  export type DateLike = Date | string;
```
Comment (patrickbenjamin1) [inline module/src/utils/dates.ts:11]: Should we support number as well for working with unix dates?

File: module/src/utils/maths.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export namespace Maths {
+  /**
+   * Perform a modulo operation that ensures that the output is always positive - javascript modulos behave unusually with negative numbers
+   * see: https://web.archive.org/web/20090717035140if_/javascript.about.com/od/problemsolving/a/modulobug.htm
+   */
+  export function positiveModulo(numerator: number, denominator: number) {
```
Comment (patrickbenjamin1) [inline module/src/utils/maths.ts:6]: beautiful

File: playground/src/views/examples/calendar.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  const maxDate = "30/06/2022"
+  const rangeToDate = initialDate
+
+  const [currentDate, setCurrentDate] = React.useState(initialDate)
+
+  const highlights = React.useMemo<ICalendarHighlight[]>(() => {
+    return [
+      { date: "02/06/2021", className: "brexit" },
+      { date: "23/06/2021", className: "means" },
+      { date: "07/07/2021", className: "breakfast" },
```
Comment (patrickbenjamin1) [inline playground/src/views/examples/calendar.tsx:17]: brexit

File: module/src/components/calendarView/calendarView.component.tsx (hunk 1 above)
Comment (jlews) [reply module/src/components/calendarView/calendarView.component.tsx:10]: Yeah that probably makes sense

File: module/src/utils/arrays.ts (hunk 2 above)
Comment (jlews) [reply module/src/utils/arrays.ts:42]: Ahh shit yeah! I tend to default to `splice` and completely forget about `slice`! I'll change it up.

File: module/src/utils/arrays.ts (hunk 3 above)
Comment (jlews) [reply module/src/utils/arrays.ts:48]: Happy to do this

File: module/src/utils/dates.ts (hunk 4 above)
Comment (jlews) [reply module/src/utils/dates.ts:11]: I'll look into it.

**Pull request merged**